buffer.go

```diff
@@ -0,0 +1,200 @@
+/*
+The Cache have RenameFrom : oldPath and RenameTo: newPath events cached.
+The time that an event was added is kept in timers cache. It's circular array.
+Every 100 miliseconds expired items are evicted from cache and if there is no match for them CREATE or DELETE event is created.
+
+
+*/
+
+package fsevents
+
+import (
+	"time"
+)
+
+const (
+	MaxCapacity = 2000
+)
+
+type TimerCache struct {
+	Values               []TimerCacheItem
+	NumberOfItems        int
+	Capacity             int
+	TimerHead, TimerTail int
+}
+type Cache struct {
+	RenameFrom map[uint64]Event
+	RenameTo   map[uint64]Event
+	timers     TimerCache
+
+	ProcessedEvents chan []Event
+	done            chan bool
+}
+type TimerCacheItem struct {
+	T  time.Time
+	ID uint64
+}
+
+func (tc *TimerCache) moveHead() {
+	tc.TimerHead = (tc.TimerHead + 1) % tc.Capacity
+	tc.NumberOfItems--
+}
+func (tc *TimerCache) Add(t time.Time, id uint64) {
+	tc.Values[tc.TimerTail] = TimerCacheItem{t, id}
+	tc.NumberOfItems++
+}
+func (tc *TimerCache) Next() {
+	tc.TimerTail = (tc.TimerTail + 1) % tc.Capacity
+}
+func (tc *TimerCache) prev() {
+	tc.TimerTail = (tc.TimerTail - 1) % tc.Capacity
+}
+func (tc *TimerCache) Head() TimerCacheItem {
+	return tc.Values[tc.TimerHead]
+}
+
+func (c *Cache) createRenameEvent(oldNameEvent, newNameEvent Event, oldNameExist, newNameExist bool) {
+	if !newNameExist && oldNameExist { //ITEM REMOVED
+		oldNameEvent.Flags = ItemRemoved
+		c.BroadcastRenameEvent(oldNameEvent)
+		return
+	} else if newNameExist && !oldNameExist { //ITEM CREATED
+		newNameEvent.Flags = ItemCreated
+	} else if newNameExist && oldNameExist { //RENAME WITH FULL INFO
+		newNameEvent.OldPath = oldNameEvent.Path
+	}
+	c.BroadcastRenameEvent(newNameEvent)
+}
+func (c *Cache) CheckForMatch(eventId uint64) (event Event, matchedEvent Event, eventExist, matchExist bool, mode string) {
+	if event, eventExist = c.RenameFrom[eventId]; eventExist {
+		delete(c.RenameFrom, event.ID)
+		matchedEvent, matchExist = c.RenameTo[event.ID+1]
+		if matchExist {
+			delete(c.RenameTo, matchedEvent.ID)
+		}
+		mode = "RENAME_FROM"
+		return
+	} else if event, eventExist = c.RenameTo[eventId]; eventExist {
+		delete(c.RenameTo, event.ID)
+		matchedEvent, matchExist = c.RenameFrom[event.ID-1]
+		if matchExist {
+			delete(c.RenameTo, matchedEvent.ID)
+		}
+		mode = "RENAME_TO"
+		return
+	}
+	return
+}
+func (c *Cache) removeHead() {
+	event, matchedEvent, eventExist, matchExist, mode := c.CheckForMatch(c.timers.Head().ID)
+	if mode == "RENAME_TO" {
+		c.createRenameEvent(matchedEvent, event, matchExist, eventExist)
+		c.timers.moveHead()
+	} else if mode == "RENAME_FROM" {
+		c.createRenameEvent(event, matchedEvent, eventExist, matchExist)
+		c.timers.moveHead()
```

Comment:
Is there a scenario where you don't `moveHead()` when calling `createRenameEvent`? If so, you could save one line and potential dev mistakes by wrapping moving the head.

Reply:
Yeah.
Used two hash maps for RENAME_TO and RENAME_FROM and the timerCache is a circular list and events are sorted by their time. `Instead of evicting events from timerCache upon finding a match , I keep checking the events in the cache every 1000ms and evict them if the above cases happen to avoid the O(n) search on every single match.`
So there is a case that the event does not exist in the `rename event list` but the timer exists.